source/language/pulses.rst

```diff
@@ -92,3 +79,217 @@ circuit. Most of the time symbols in the ``defcal`` table will also have
+requested phase change of type angle. The exact precision of the implemented
+phase change will vary depending on hardware support.
+
+.. code-block:: none
+
+   // Shift phase of qubit 0 by pi/4, eg. an rz gate with angle -pi/4
+   shift_phase drive(%0), pi/4;
+
+   // Define a calibration for the rz gate on all physical qubits
+   defcal rz(angle[20]:theta) %q {
+     shift_phase drive(%q), -theta;
+   }
+
+Shift Frequency Instruction
+---------------------------
+
+A ``shift_freq`` instruction schedules a frequency advance for all the following
+pulses on that channel. This is useful for defining spectroscopy experiments.
+
+The ``shift_freq`` instruction takes two parameters, a channel and the
+requested frequency change of type float in units of GHz.
+
+Here's an example qubit spectroscopy experiment. Note that the starting
+frequency will be defined somewhere outside OpenQASM.
+
+.. code-block:: none
+
+   qubit q;
+
+   const shots = 1000;
+   const dfreq = 0.001; # 1MHz per point
+   const points = 50; # Sweep over 50MHz
+
+   complex[float[32]] iq, average;
+   complex[float[32]] output[points];
+
+   for p in [0 : points-1] {
+     average = 0;
+     for i in [0 : shots-1] {
+       // Assumes suitable calibrations for reset, x, and measure_iq
+       reset q;
+       x q;
+       measure_iq q -> iq;
+
+       average = (average * i + iq) / (i + 1);
+     }
+     shift q;
+     output[p] = average;
+   }
+
+   defcal shift %q {
+     shift_freq drive(%q) dfreq;
+   }
+
+Capture Instruction
+-------------------
+
+Acquisition is scheduled by a ``capture`` instruction.
+
+The ``capture`` instruction takes two parameters, a channel and the filter to
+apply to the returned signal. The length of the filter in dt will determine how
+long the capture channel is open. The ``capture`` instruction returns the dot
+product of the measured IQ values and the filter. (Note that the filter is
+sometimes referred to as "kernel" in other contexts, but this is not related in
+any way to the ``kernel`` instruction in OpenQASM).
+
+.. code-block:: none
+    
+   complex[float[32]] filter = [1, 1, 1, 1, 1];
+   // Capture for 5 samples
+   iq = capture acquire(%0), filter;
+
+Specifying a full list of samples for real-life filters can be unwieldy, so we
+include several built-in filter functions as well. Note that these return the
+same type as pulse shape functions and therefore either can be used for pulse 
+shapes and filters.
+
+.. code-block:: none
+
+   // Define a boxcar (aka. constant) filter of length l
+   boxcar(l:length)
+
+Timing
+------
+
+Each channel maintains its own "clock". When a pulse is played the clock for 
+that channel advances by the number of samples in the pulse. The same is true
+for output channels based on the length of the capture filter. Pulses on a
+single channel cannot be played simultaneously, although pulses on multiple
+channels for the same qubit can.
```

Comment:
Sometimes it is necessary to apply crosstalk compensation pulses in our systems that effectively occupy the same hardware channel at once such as in the case of a geometric gate or stark tone. While these could be captured by constructing a complex signal envelope it may be nice to consider a more graceful way of expressing this. I am ok pushing this down the road.